pkg/kubecost/allocation.go

```diff
@@ -2405,3 +2405,67 @@ func (asr *AllocationSetRange) Window() Window {
 
 	return NewWindow(&start, &end)
 }
+
+// Start returns the earliest start of all Allocations in the AllocationSetRange.
+// It returns an error if there are no allocations.
+func (asr *AllocationSetRange) Start() (time.Time, error) {
+	start := time.Date(1970, 1, 1, 0, 0, 0, 0, time.UTC)
+	firstStartNotSet := true
+	asr.Each(func(i int, as *AllocationSet) {
+		as.Each(func(s string, a *Allocation) {
+			if firstStartNotSet {
+				start = a.Start
+				firstStartNotSet = false
+			}
+			if a.Start.Before(start) {
+				start = a.Start
+			}
+		})
+	})
+
+	if firstStartNotSet {
+		return start, fmt.Errorf("had no data to compute a start from")
+	}
+
+	return start, nil
+}
+
+// End returns the lateste end of all Allocations in the AllocationSetRange.
+// It returns an error if there are no allocations.
+func (asr *AllocationSetRange) End() (time.Time, error) {
+	end := time.Date(1970, 1, 1, 0, 0, 0, 0, time.UTC)
+	firstEndNotSet := true
+	asr.Each(func(i int, as *AllocationSet) {
+		as.Each(func(s string, a *Allocation) {
+			if firstEndNotSet {
+				end = a.End
+				firstEndNotSet = false
+			}
+			if a.End.After(end) {
+				end = a.End
+			}
+		})
+	})
+
+	if firstEndNotSet {
+		return end, fmt.Errorf("had no data to compute an end from")
+	}
+
+	return end, nil
+}
+
+// Minutes returns the duration, in minutes, between the earliest start
+// and the latest end of all allocations in the AllocationSetRange.
+func (asr *AllocationSetRange) Minutes() (float64, error) {
+	start, err := asr.Start()
+	if err != nil {
+		return 0, fmt.Errorf("failed to calculate start: %s", err)
+	}
+	end, err := asr.End()
+	if err != nil {
+		return 0, fmt.Errorf("failed to calculate end: %s", err)
+	}
```

Comment:
What do you think about just returning `0` if there are no allocations?

Reply:
Because the methods this calls (`Start` and `End`) can error, I think I have to propagate that error because otherwise I make an assumption that the only error states `Start` and `End` will be in are "no allocations."

OTOH I do think it is entirely reasonable for 0 to be returned here in the event of an error on start and end, as the value itself isn't actually invalid and callers currently have no need of knowing if the 0 is the result of an error or not. I'll change it.